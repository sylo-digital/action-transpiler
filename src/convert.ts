import { replaceAdvancedEmbed } from "./conversions/replace-advancedembed";
import { replaceArgs } from "./conversions/replace-args";
import { replaceBrackets } from "./conversions/replace-brackets";
import { replaceChoose } from "./conversions/replace-choose";
import { replaceCommandTags } from "./conversions/replace-command-tags";
import { replaceCreatedAt } from "./conversions/replace-createdAt";
import { replaceGeneric } from "./conversions/replace-generic";
import { replaceLowerUpper } from "./conversions/replace-lower-upper";
import { replaceMessageTimestamp } from "./conversions/replace-message-timestamp";
import { replaceNotesWithComments } from "./conversions/replace-notes-with-comments";
import { replaceRandstr } from "./conversions/replace-randstr";
import { replaceVariables } from "./conversions/replace-variables";
import { treeToText } from "./helpers/tree-to-text";
import { Lexer } from "./lexer";
import { Converter, ScriptOptions } from "./types";

const converters: Converter[] = [
  replaceGeneric,
  replaceArgs,
  replaceNotesWithComments,
  replaceBrackets,
  replaceVariables,
  replaceCommandTags,
  replaceMessageTimestamp,
  replaceCreatedAt,
  replaceChoose,
  replaceLowerUpper,
  replaceRandstr,
  replaceAdvancedEmbed, // must be last
];

export function convert(script: string) {
  const options: ScriptOptions = {
    appendAfter: [],
    appendBefore: [],
    source: script,
    warnings: [],
    errors: [],
    injectInputOption: false,
  };

  const lexer = new Lexer(script);
  const tree = lexer.parse();
  for (const converter of converters) {
    converter(tree, options);
  }

  let text = treeToText(tree);
  if (options.warnings[0]) {
    text = `${options.warnings.map((warning) => `// ${warning}`).join("\n")}\n${text}`;
  }

  if (options.appendAfter) {
    text = options.appendBefore.join("\n") + "\n" + text;
  }

  if (options.appendAfter[0]) {
    text = text + "\n" + options.appendAfter.join("\n");
  }

  return {
    output: text.trim(),
    warnings: options.warnings,
    errors: options.errors,
    injectInputOption: options.injectInputOption,
  };
}
