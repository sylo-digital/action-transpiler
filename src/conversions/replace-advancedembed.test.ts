import dedent from "dedent";
import { convert } from "../convert";

it("should convert nested tags", () => {
  // because the converter returns a string with the stringified tag, nested tags
  // are missed unless the ae conversion is run last once they've already been replaced.
  const result = convert(`{a!ae;--description="{perget;name}"}`).output;
  expect(result).toBe('{responder.embed;{{"description":"{store.get;name}"}}}');
});

it("should convert nested tags", () => {
  const script = dedent`
    {channel.send;
      {a!ae;
      --edit="725492484326686750";
      --color="#a3d5ae";
      --field1name="Pick One";
      --field1value=":spider: - {role.mention;702325354748182548}\n:scorpion: - {role.mention;702325340105736303}";
      --field1inline;
      --title="Location Roles";
      --description="React on the message to chose your Hemisphere:"};723698264511086624}
  `;

  expect(convert(script).output).toMatchSnapshot();
});

it("should convert embeds with no description", () => {
  const script = dedent`
      {a!ae;
        --color="#a3d5ae";
        --title="{$legacy_args}";
        --footer="{user.tag}";
        --footer-icon="{user.avatarURL}";
        --timestamp
      }
  `;

  expect(convert(script).output).toMatchSnapshot();
});
