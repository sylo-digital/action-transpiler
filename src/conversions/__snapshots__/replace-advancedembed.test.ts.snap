// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should convert embeds with no description 1`] = `
"{responder.embed;{{
  \\"color\\": 10737070,
  \\"title\\": \\"{$legacy_args}\\",
  \\"footer\\": {
    \\"icon_url\\": \\"{user.avatar}\\",
    \\"text\\": \\"{user.tag}\\"
  },
  \\"timestamp\\": \\"2022-04-03T15:49:18.989Z\\"
}}}"
`;

exports[`should convert nested tags 1`] = `
"// This function has been injected to replace functionality of a legacy {channel.send} tag.
[#function;legacy_channel_send;{=message};{=channel}]
    {responder.text;{$message}} // Add the text to the responder
    {=channelId;{or;{$channel};{channel.id}}} // Get the channel we're sending to
    {responder.channel;{$channelId}} // Configure the responder to send the message to the channel
    {responder.send} // Send the message
[/function]
{legacy_channel_send;
  {responder.embed;{{
  \\"color\\": 10737070,
  \\"title\\": \\"Location Roles\\",
  \\"description\\": \\"React on the message to chose your Hemisphere:\\",
  \\"fields\\": [
    {
      \\"name\\": \\"Pick One\\",
      \\"inline\\": true,
      \\"value\\": \\":spider: - {role.mention;702325354748182548}\\\\n:scorpion: - {role.mention;702325340105736303}\\"
    }
  ]
}}};723698264511086624}"
`;
